### ISSUE TITLE
recode_road_class

### ISSUE THREAD

**Issue by hulsiejames** [udsleeds/openinfra#78]
It seems (at least in Leeds - likely elsewhere) that some residential roads are listed as `highway="service"` 

![](https://i.imgur.com/0Xku3SJ.png)

Need to look into adding this to recode_road_class function. 

Hopefully there is a logcial way we can distinguish between genuine service roads, and residential roads labelled as service.

**Comment by hulsiejames** [comment]
Update: 

Back Hamilton Ave is marked as `highway=service`, which is actually in convention with the documentation on [highway=service](https://wiki.openstreetmap.org/wiki/Tag:highway%3Dservice)
> This is also commonly used for access to parking, driveways, and alleys.

Of which Back Hamilton Ave is (also the other roads inbetween gardens, see Back Grange Ave above.)

Visualising on street view we see:
![](https://i.imgur.com/oFyFE7F.jpg)

The next road up (front house facing) is classified as a local road, and mapped as `highway=residential` appropriately is Hamilton Avenue:
![](https://i.imgur.com/LPztadx.jpg)

This is captured by the current recode_road_class function: 
![](https://i.imgur.com/YmBX57d.png)

The question: Is this okay functionality? - do we need to be capturing such back alleys/parking routes as these are not likely to be prioritised in terms of infrastructure development.

If we do wish to try and capture such back alleys - how? (would need some thought). 
Looking at the intial alley road tags (for Back Hamilton Ave - same for all Back Aves here):
![](https://i.imgur.com/Q77lgq5.png)
I don't see an immediate way we could classify them as such? 
One thought could be seeing if these `highway=service` ways intersect or connect to a way tagged `highway=residential` as the service road is likely also used primarily by residents, as in the examples here. 

Personally I'm not sure such information is useful to local authorities who will be focused on more important aspects, but that is just an opinion!

**Comment by Robinlovelace** [comment]
> One thought could be seeing if these `highway=service` ways intersect or connect to a way tagged `highway=residential` as the service road is likely also used primarily by residents, as in the examples here.

That is a good idea :+1: there may be other ways of doing it but this is a solid starter for 10.

**Comment by Robinlovelace** [comment]
> I will try and create a plot of `highway=service` and corresponding `width=*` to see the distribution, though I expect not many will have width added!

I wasn't referring to `width=` tags in OSM. I meant are some of them larger non-residential roads in reality.

**Comment by hulsiejames** [comment]
I see - I can't think of a way of identifying such misclassified roads at this moment, apart from maybe a width sense check.

I.e. if a way has `highway=service` and a `width` of 5 meters or greater it could likely be a missclassified non-residential road

**Comment by Robinlovelace** [comment]
Looking at overpass turbo it seems most service roads are eminently cyclable: 

![image](https://user-images.githubusercontent.com/1825120/177985384-8f011372-2be6-461c-8104-3227fc0245c4.png)

Source: https://overpass-turbo.eu/

**Comment by Robinlovelace** [comment]
Although some of them are definitely NOT cyclable. So basing it on whether or not they connect with residential (and cyclways/other active travel ways) is my suggestion.

![image](https://user-images.githubusercontent.com/1825120/177985852-ef54c27b-8092-488e-a3ad-6b6a3ad78e0a.png)

**Comment by Robinlovelace** [comment]
Source: https://www.openstreetmap.org/way/673677250

Does access:no cover it?

**Comment by hulsiejames** [comment]
> Source: https://www.openstreetmap.org/way/673677250
Does access:no cover it?

And access:private

Source: https://www.openstreetmap.org/edit?way=673677250#map=17/53.73074/-1.51185

**Comment by Robinlovelace** [comment]
Does that mean that it will be excluded from the cycle network and by extension the openinfra packs? I should probably check the code, but quicker to ask! I guess the key questions are:

- Are there any false positives from highway:service? 
- How should they be classified (residential if they intersect with residential roads being a good default, what about service ways that do not?)?

**Comment by Robinlovelace** [comment]
> so perhaps there could be a check for that?

Yes, good sanity checking and worth excluding service roads that are not on active travel networks.

**Comment by Robinlovelace** [comment]
(It's still good to do the check in this repo IMO and not rely too much on external dependencies; starting from the full raw OSM data and keeping all the rules in one place could be good for documentation and transparency to people reading the code and trying to get their head round the outputs.)

**Comment by hulsiejames** [comment]
> Does that mean that it will be excluded from the cycle network and by extension the openinfra packs? I should probably check the code, but quicker to ask! I guess the key questions are:

Currently, any way with `highway="service"` is **not** included by `recode_road_class`.

The "cycleway" option in `recode_road_class` is only matching to ways with `highway="cycleway"`, rather than indicating a cycleable network (this will be captured by `oi_active_cycle` column, decribed below). 

Origionally Chan and Cooper defined this class as "Traffic Free Roads", however as `highway="cycleway" indicates a seperate way for cyclists, it is not a road at all thus I changed the description to "cycleway"


The next function I will make `oi_active_travel` which will include two columns `oi_active_walk` & `oi_active_cycle` which will contain "yes", "no" values based on whether an OSM way is included by `oe_get_network(type="walking")` OR `(type="cycling")`

On classifying such `highway="service"` ways
>How should they be classified (residential if they intersect with residential roads being a good default, what about service ways that do not?)?

I agree classifying ways with highway="service" as residential when intesecting with other ways labelled as  `highway="residential"` in the `recode_road_class` function. 

There may also be a way to identifty highway="service" roads that lead into industrial estates - these should probably not be counted as local/residential roads but as industrial estate? bearing in mind we are creating more road descriptions we dont want to overwhelm

**Comment by Robinlovelace** [comment]
> The next function I will make `oi_active_travel` which will include two columns `oi_active_walk` & `oi_active_cycle` which will contain "yes", "no" values based on whether an OSM way is included by `oe_get_network(type="walking")` OR `(type="cycling")`

Based on the concept of modularity, I suggest first creating/documenting two separate functions and then combining them.

**Comment by Robinlovelace** [comment]
> There may also be a way to identifty highway="service" roads that lead into industrial estates - these should probably not be counted as local/residential roads but as industrial estate? bearing in mind we are creating more road descriptions we dont want to overwhelm

Quickfire idea on this: add an 'other' category. There is a diversity of roads and a balance between simplicity (good) and the dataset's ability to describe the real world in a decent amount of detail (also good). We don't need to stick religiously to Chan and Cooper's definition, we can build on it so I suggest separating out cycleway from traffic free roads. Putting this together my current thinking is:

- New 'other' category that may not fit into the others
- Add cycleway as a distinct category that is worthy of documenting in open transport infrastructure data packs
- Add highway=foot and other walking-only ways (e.g. steps) as another category
- Keep the car-free category buy maybe rename it and redefine it as ways on which cars can go but on which there is no through traffic

That final point links to long conversations about how to define LTNs based on OSM data alone: https://github.com/a-b-street/abstreet/discussions/770

**Comment by Robinlovelace** [comment]
FYI we can could build on CycleStreets' definition of LTNs and their API: https://www.cyclestreets.net/api/v2/advocacydata.ltns/

See here for an R function that gets ways classified by LTN status from CycleStreets. A question here is whether LTNs are defined based on infrastructure e.g. planters preventing rat running and sign-posts saying there's no through traffic or regulation (my thinking: regulation without infrastructure is unusual and weak and that given the importance of no through traffic roads aka LTNs it's worth representing them in our data packs).

**Comment by hulsiejames** [comment]
> Quickfire idea on this: add an 'other' category. There is a diversity of roads and a balance between simplicity (good) and the dataset's ability to describe the real world in a decent amount of detail (also good). We don't need to stick religiously to Chan and Cooper's definition, we can build on it so I suggest separating out cycleway from traffic free roads. Putting this together my current thinking is:
> 
> * New 'other' category that may not fit into the others
> * Add cycleway as a distinct category that is worthy of documenting in open transport infrastructure data packs
> * Add highway=foot and other walking-only ways (e.g. steps) as another category
> * Keep the car-free category buy maybe rename it and redefine it as ways on which cars can go but on which there is no through traffic
> 
> That final point links to long conversations about how to define LTNs based on OSM data alone: [a-b-street/abstreet#770](https://github.com/a-b-street/abstreet/discussions/770)

Just to clarify the car-free category would be it's own column within the datapack rather than part of the functionality of `recode_road_class` - could be part of a function that calls the cyclestreet API or determins rat-runs/through roads/LTNs etc.?

I like this - current idea for recode_road_class descriptions is for the follwoing: 
 - Dual Carriageways (R & NR - distinction added later - highway="trunk" or "trunk_link")
 - Motorways (highway="motorway" or "motoway_link")
 - Primary Roads (highway="primary" or "primary_link")
 - Secondary Roads (highway="secondary" or "secondary_link")
 - Tertiary Roads (highway="tertiary" or "tertiary_link")
 - Residential/Local Roads (highway="residential" or "living_street" or "unclassified") ( *Looking to include highway="service" if attached to a way containing highway="residential"*)
 - Other (Any road not captured by the above categories)

Is your plan to have Cycleways (as a piece of infrastrcutre) included in the `oi_highways column`, the output of `recode_road_class` - or to have a seperate column entirely that is just called `oi_cycleway` which has values "yes" if the feature is a cycleway? 

The above will be in addition to `oi_active_cycle` which will be the cycling network returned by osmextract `oe_get_network(type="walking")`, not just seperate ways designated for cyclists

Similarly with walking the walking point, should I create two instances? one say `oi_active_walk` which is the network returned from osmextract `oe_get_network(type="walking")` and a second, say `oi_walk` that returns specific ways designated for pedestrian traffic only?

**Comment by Robinlovelace** [comment]
Lots of good questions in there. One to discuss on the phone and probably with practitioners like Brian Deegan, Kit Allwinter and Catriona Swanson!
